### ISSUE TITLE
Evaluation accuracy is either nan or 0 for most of the classes

### ISSUE THREAD

**Issue by askgitqa** [open-mmlab/mmsegmentation#2021]
Hi,
I am using `PSPNET` for 38 class segmentation but only 6-7 classes have proper segmentation accuracy rest are either 0 or nan. Here is my config file.
```
from mmseg.apis import set_random_seed
from mmseg.utils import get_device
from mmcv import Config
from mmseg.apis import set_random_seed
from mmseg.utils import get_device

config_file = 'configs/pspnet/pspnet_r18b-d8_512x1024_80k_cityscapes.py'
checkpoint_file = 'checkpoints/pspnet_r18b-d8_512x1024_80k_cityscapes_20201226_063116-26928a60.pth'

cfg = Config.fromfile(config_file)
cfg.load_from = checkpoint_file

# Since we use only one GPU, BN is used instead of SyncBN
cfg.norm_cfg = dict(type='BN', requires_grad=True)
cfg.model.backbone.norm_cfg = cfg.norm_cfg
cfg.model.decode_head.norm_cfg = cfg.norm_cfg
cfg.model.auxiliary_head.norm_cfg = cfg.norm_cfg

# modify num classes of the model in decode/auxiliary head
cfg.model.decode_head.num_classes = len(classes)
cfg.model.auxiliary_head.num_classes = len(classes)

# Modify dataset type and path
cfg.dataset_type = 'StanfordBackgroundDataset'
cfg.data_root = data_root

cfg.data.samples_per_gpu = 8
cfg.data.workers_per_gpu=1

cfg.img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)

crop_size = (500, 500)
cfg.crop_size = crop_size
cfg.train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations'),
    dict(type='Resize', img_scale=crop_size, ratio_range=(0.5, 2.0)),
    dict(type='RandomCrop', crop_size=cfg.crop_size, cat_max_ratio=0.75),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='PhotoMetricDistortion'),
    dict(type='Normalize', **cfg.img_norm_cfg),
    dict(type='Pad', size=cfg.crop_size, pad_val=0, seg_pad_val=255),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg']),
]

cfg.test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=crop_size,
        # img_ratios=[0.5, 0.75, 1.0, 1.25, 1.5, 1.75],
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **cfg.img_norm_cfg),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]


cfg.data.train.type = cfg.dataset_type
cfg.data.train.data_root = cfg.data_root
cfg.data.train.img_dir = img_dir
cfg.data.train.ann_dir = ann_dir
cfg.data.train.pipeline = cfg.train_pipeline
cfg.data.train.split = 'splits/train.txt'

cfg.data.val.type = cfg.dataset_type
cfg.data.val.data_root = cfg.data_root
cfg.data.val.img_dir = img_dir
cfg.data.val.ann_dir = ann_dir
cfg.data.val.pipeline = cfg.test_pipeline
cfg.data.val.split = 'splits/val.txt'

cfg.data.test.type = cfg.dataset_type
cfg.data.test.data_root = cfg.data_root
cfg.data.test.img_dir = img_dir
cfg.data.test.ann_dir = ann_dir
cfg.data.test.pipeline = cfg.test_pipeline
cfg.data.test.split = 'splits/val.txt'



# Set up working dir to save files and logs.
cfg.work_dir = './work_dirs/tutorial'

cfg.runner.max_iters = 200
cfg.log_config.interval = 10
cfg.evaluation.interval = 200
cfg.checkpoint_config.interval = 200

# Set seed to facitate reproducing the result
cfg.seed = 0
set_random_seed(0, deterministic=False)
cfg.gpu_ids = range(1)
cfg.device = get_device()

# Let's have a look at the final config used for training
# print(f'Config:\n{cfg.pretty_text}')
```
 
Here is the complete training log... 
```
 -------------------- 
2022-09-03 09:09:38,240 - mmseg - INFO - Iter [10/200]	lr: 9.598e-03, eta: 0:25:39, time: 8.103, data_time: 7.600, memory: 4370, decode.loss_ce: 2.3194, decode.acc_seg: 42.4490, aux.loss_ce: 1.1253, aux.acc_seg: 14.5177, loss: 3.4447
2022-09-03 09:10:25,394 - mmseg - INFO - Iter [20/200]	lr: 9.149e-03, eta: 0:19:13, time: 4.715, data_time: 4.249, memory: 4370, decode.loss_ce: 1.0370, decode.acc_seg: 71.4401, aux.loss_ce: 0.7380, aux.acc_seg: 67.0020, loss: 1.7750
2022-09-03 09:11:12,140 - mmseg - INFO - Iter [30/200]	lr: 8.698e-03, eta: 0:16:31, time: 4.673, data_time: 4.199, memory: 4370, decode.loss_ce: 0.6922, decode.acc_seg: 78.1816, aux.loss_ce: 0.4196, aux.acc_seg: 74.9002, loss: 1.1118
2022-09-03 09:11:58,764 - mmseg - INFO - Iter [40/200]	lr: 8.244e-03, eta: 0:14:46, time: 4.661, data_time: 4.174, memory: 4370, decode.loss_ce: 0.8181, decode.acc_seg: 73.8185, aux.loss_ce: 0.4262, aux.acc_seg: 69.1969, loss: 1.2444
2022-09-03 09:12:45,273 - mmseg - INFO - Iter [50/200]	lr: 7.788e-03, eta: 0:13:24, time: 4.650, data_time: 4.151, memory: 4370, decode.loss_ce: 0.7979, decode.acc_seg: 73.3402, aux.loss_ce: 0.4034, aux.acc_seg: 69.5064, loss: 1.2013
2022-09-03 09:13:32,245 - mmseg - INFO - Iter [60/200]	lr: 7.328e-03, eta: 0:12:14, time: 4.698, data_time: 4.206, memory: 4370, decode.loss_ce: 0.7810, decode.acc_seg: 76.6029, aux.loss_ce: 0.3801, aux.acc_seg: 72.2736, loss: 1.1611
2022-09-03 09:14:19,012 - mmseg - INFO - Iter [70/200]	lr: 6.865e-03, eta: 0:11:11, time: 4.676, data_time: 4.177, memory: 4370, decode.loss_ce: 0.6710, decode.acc_seg: 77.0327, aux.loss_ce: 0.3365, aux.acc_seg: 73.9599, loss: 1.0076
2022-09-03 09:15:05,473 - mmseg - INFO - Iter [80/200]	lr: 6.398e-03, eta: 0:10:12, time: 4.646, data_time: 4.148, memory: 4370, decode.loss_ce: 0.5764, decode.acc_seg: 79.5449, aux.loss_ce: 0.3193, aux.acc_seg: 74.1004, loss: 0.8957
2022-09-03 09:15:52,048 - mmseg - INFO - Iter [90/200]	lr: 5.928e-03, eta: 0:09:15, time: 4.657, data_time: 4.150, memory: 4370, decode.loss_ce: 0.6571, decode.acc_seg: 77.8865, aux.loss_ce: 0.3320, aux.acc_seg: 73.6684, loss: 0.9891
2022-09-03 09:16:11,885 - mmseg - INFO - Iter [100/200]	lr: 5.453e-03, eta: 0:07:54, time: 1.983, data_time: 1.057, memory: 4370, decode.loss_ce: 0.5928, decode.acc_seg: 81.0651, aux.loss_ce: 0.3125, aux.acc_seg: 76.8401, loss: 0.9052
2022-09-03 09:16:23,170 - mmseg - INFO - Iter [110/200]	lr: 4.974e-03, eta: 0:06:37, time: 1.129, data_time: 0.078, memory: 4370, decode.loss_ce: 0.5334, decode.acc_seg: 82.5853, aux.loss_ce: 0.2892, aux.acc_seg: 77.7679, loss: 0.8225
2022-09-03 09:16:34,364 - mmseg - INFO - Iter [120/200]	lr: 4.489e-03, eta: 0:05:31, time: 1.120, data_time: 0.076, memory: 4370, decode.loss_ce: 0.6577, decode.acc_seg: 79.4756, aux.loss_ce: 0.3432, aux.acc_seg: 74.5075, loss: 1.0009
2022-09-03 09:16:45,427 - mmseg - INFO - Iter [130/200]	lr: 3.998e-03, eta: 0:04:33, time: 1.107, data_time: 0.076, memory: 4370, decode.loss_ce: 0.5568, decode.acc_seg: 80.7693, aux.loss_ce: 0.3077, aux.acc_seg: 75.3771, loss: 0.8645
2022-09-03 09:16:56,400 - mmseg - INFO - Iter [140/200]	lr: 3.500e-03, eta: 0:03:42, time: 1.097, data_time: 0.076, memory: 4370, decode.loss_ce: 0.6020, decode.acc_seg: 79.4450, aux.loss_ce: 0.3113, aux.acc_seg: 74.7616, loss: 0.9134
2022-09-03 09:17:07,359 - mmseg - INFO - Iter [150/200]	lr: 2.994e-03, eta: 0:02:56, time: 1.096, data_time: 0.075, memory: 4370, decode.loss_ce: 0.5480, decode.acc_seg: 81.1858, aux.loss_ce: 0.2947, aux.acc_seg: 75.2111, loss: 0.8427
2022-09-03 09:17:18,374 - mmseg - INFO - Iter [160/200]	lr: 2.478e-03, eta: 0:02:15, time: 1.101, data_time: 0.075, memory: 4370, decode.loss_ce: 0.5955, decode.acc_seg: 80.8860, aux.loss_ce: 0.3034, aux.acc_seg: 77.0725, loss: 0.8989
2022-09-03 09:17:29,457 - mmseg - INFO - Iter [170/200]	lr: 1.949e-03, eta: 0:01:37, time: 1.108, data_time: 0.076, memory: 4370, decode.loss_ce: 0.5276, decode.acc_seg: 82.8717, aux.loss_ce: 0.2933, aux.acc_seg: 77.0121, loss: 0.8209
2022-09-03 09:17:40,564 - mmseg - INFO - Iter [180/200]	lr: 1.402e-03, eta: 0:01:02, time: 1.111, data_time: 0.076, memory: 4370, decode.loss_ce: 0.5698, decode.acc_seg: 81.9036, aux.loss_ce: 0.3128, aux.acc_seg: 76.0378, loss: 0.8827
2022-09-03 09:17:53,321 - mmseg - INFO - Iter [190/200]	lr: 8.277e-04, eta: 0:00:30, time: 1.275, data_time: 0.304, memory: 4370, decode.loss_ce: 0.5630, decode.acc_seg: 81.6323, aux.loss_ce: 0.2973, aux.acc_seg: 76.4273, loss: 0.8603
2022-09-03 09:18:03,757 - mmseg - INFO - Saving checkpoint at 200 iterations
2022-09-03 09:18:06,010 - mmseg - INFO - Iter [200/200]	lr: 1.841e-04, eta: 0:00:00, time: 1.337, data_time: 0.077, memory: 4370, decode.loss_ce: 0.5258, decode.acc_seg: 83.2014, aux.loss_ce: 0.2871, aux.acc_seg: 77.1780, loss: 0.8129
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 186/186, 1.8 task/s, elapsed: 104s, ETA:     0s2022-09-03 09:19:50,051 - mmseg - INFO - per class results:
2022-09-03 09:19:50,056 - mmseg - INFO - 
+------------+-------+-------+
|   Class    |  IoU  |  Acc  |
+------------+-------+-------+
| background | 82.57 | 89.98 |
|  class1    |  0.0  |  0.0  |
|  class2    |  nan  |  nan  |
|  class3    |  0.0  |  0.0  |
|  class4    |  0.0  |  0.0  |
|  class5    |  0.0  |  0.0  |
|  class6    |  0.0  |  0.0  |
|  class7    | 19.22 | 35.24 |
|  class8    |  0.0  |  0.0  |
|  class9    |  0.0  |  0.0  |
|  class10   |  0.0  |  0.0  |
|  class11   |  0.0  |  0.0  |
|  class12   |  0.0  |  0.0  |
|  class13   |  nan  |  nan  |
|  class14   |  0.0  |  0.0  |
|  class15   |  nan  |  nan  |
|  class16   |  0.0  |  0.0  |
|  class17   |  0.0  |  0.0  |
|  class18   |  nan  |  nan  |
|  class19   |  0.0  |  0.0  |
|  class20   |  0.0  |  0.0  |
|  class21   |  0.0  |  0.0  |
|  class22   |  0.0  |  0.0  |
|  class23   |  nan  |  nan  |
|  class24   | 68.92 | 94.76 |
|  class25   | 70.08 | 90.33 |
|  class26   |  0.0  |  0.0  |
|  class27   |  0.0  |  0.0  |
|  class28   |  0.0  |  0.0  |
|  class29   |  nan  |  nan  |
|  class30   |  nan  |  nan  |
|  class31   |  0.0  |  0.0  |
|  class32   |  0.0  |  0.0  |
|  class33   |  0.0  |  0.0  |
|  class34   |  nan  |  nan  |
|  class35   |  0.0  |  0.0  |
|  class36   |  0.0  |  0.0  |
|  class37   |  nan  |  nan  |
+------------+-------+-------+
```

**Comment by xiexinch** [comment]
Hi @askgitqa,
What is the size of your dataset? Because your dataset has a relatively large number of classes, training only a few iters may not work well, I think you can train more iters.

**Comment by askgitqa** [comment]
Hi Thanks for suggestion. Log shows that training accuracy has reached around 75% and there after it started fluctuation between 65 to 75 so I don't think training it for more iter would help. Kindly suggest. 
There are classes which have only 40-50 instances out of my 10k images and these classes are either 0 or nan so I think data imbalance is issue. Any idea how can I handle it in code if possible since dataset is fixed so can't change dataset.

**Comment by xiexinch** [comment]
Hi @askgitqa,
You may refer to our documentation.
https://mmsegmentation.readthedocs.io/en/latest/tutorials/training_tricks.html#class-balanced-loss
